src/main/java/QueryServlet.java

```diff
@@ -218,12 +244,134 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
+              ImmutableList.of(rightSubtract, rightSubtract, rightSubtract, rightSubtract))
+          .build();
+
+  private String reformatDataArray(String data, String dataIdentifier, boolean isCountyQuery)
+      throws InvalidObjectException {
+    if (!dataRowToReformatFunction.containsKey(dataIdentifier)) {
+      return data; // This data doesn't need reformatting
+    }
+    if ((dataRowToReformatFunction.get(dataIdentifier).size() + 1)
+        != dataIdentifier.split(",").length) {
+      // We should have enough functions listed to apply them to each group
+      // of two rows - (a, b) and then (their product, c), etc.
+      throw new InvalidObjectException(
+          "There aren't the right number of functions to apply to these data columns.");
+    }
+
+    // Convert from the census API's JSON string to a Java matrix
+    Gson gson = new Gson();
+    Type dataArrayType = new TypeToken<ArrayList<List<String>>>() {}.getType();
+    ArrayList<List<String>> originalDataArray = gson.fromJson(data, dataArrayType);
+    ArrayList<List<String>> newDataArray = new ArrayList<List<String>>();
+
+    // Create a new matrix to contain the reformatted data. Add in the the first column, which is
+    // the string identifier of the location, and the second column, the first numerical value
+    for (int i = 0; i < originalDataArray.size(); i++) {
+      List<String> originalDataRow = originalDataArray.get(i);
+      List<String> newDataRow = new ArrayList<String>();
+      newDataRow.add(originalDataRow.get(0));
+      newDataRow.add(originalDataRow.get(1));
+      newDataArray.add(newDataRow);
+    }
+
+    // Combine the original numbers, skipping the header row, by iterating over the functions
+    // given for these data columns. For each row, we successively combine the next datapoint
+    // with the previous result, finally resulting in one single fully-combined data point per row.
+    List<BiFunction> numberCombiners = dataRowToReformatFunction.get(dataIdentifier);
+    for (int i = 0; i < numberCombiners.size(); i++) { // Moving across columns combining numbers
+      BiFunction numberCombiner = numberCombiners.get(i);
+      for (int j = 1; j < originalDataArray.size(); j++) { // Moving down each row
+        List<String> originalDataRow = originalDataArray.get(j);
+        List<String> newDataRow = newDataArray.get(j);
+        if (newDataRow.get(1) == null || newDataRow.get(1).startsWith("-")) {
+          // Sometimes the census API returns missing or negative numbers by mistake;
+          // these get cleaned up in the JavaScript
+          newDataRow.set(1, "-1");
+        } else {
+          try {
+            int newValue =
+                (int)
+                    numberCombiner.apply(
+                        Float.parseFloat(newDataRow.get(1)), // Always replacing previous value
+                        Float.parseFloat(
+                            originalDataRow.get(i + 2))); // Moving along list of original values,
```

Comment:
Maybe dataRowToReformatFunction should be Map<String, List<Pair<BiFunction,Integer>>> where the Integer value indicates which column from the original data to use for this operation. then you can use that value here instead of i+2

Reply:
I'm a bit confused by what you mean, sorry. The value of i changes as we iterate through the list of BiFunctions - when we're at the first one, it's zero, then for the second one it's 1, etc. So as we go through all the functions, we also move across all the columns of the original data; we just skip the first two.

Reply:
I mean that you are using the position of the BiFunction in the list to determine the column it applies (with a +2). Instead you could explicitly specify that value along with the BiFunction, it would look something like 

put("S0201_124E,S0201_125E,S0201_126E,S0201_119E", ImmutableList.of(<add,3>, <add,4>, <percent,5>))

which translates to add data from column 3, then add data from column 4 and then use value in column 5 to calculate percentage. 

Currently since you always have them in order both these approaches work the same but image a case where you have to do the percent first and then the two adds then you would just rewrite this as 

put("S0201_124E,S0201_125E,S0201_126E,S0201_119E", ImmutableList.of(<percent,5>, <add,3>, <add,4>))

but your current approach has no simple way of supporting this. 

I don't want you to spend too much time on this so I am ok with keeping as is if you find this would be too much work.

Reply:
Ah, I understand now, thank you. I'm not opposed to implementing this but I do think it adds a good amount of complexity. I originally purposely rearranged all the lists of data so that they could proceed in order, as that seemed simpler to me at the time -- it makes no difference to the census API what order you request them in so in your example, you could just rearrange the list to get 3, 4, 5 in order again.